src/modules/csm/csm.py

```diff
@@ -422,24 +421,37 @@ def get_accumulated_strikes(self, cid: CID, root: HexBytes) -> Iterator[StrikesT
         if tree.root != root:
             raise ValueError("Unexpected tree root got from IPFS dump")
 
-        for v in tree.tree.values:
-            yield v["value"]
+        return {(no_id, pubkey): strikes for no_id, pubkey, strikes in tree.values}
 
-    def _process_raw_strikes(
+    def _merge_strikes(
         self,
-        strikes: dict[StrikesValidator, list[int]],
-        blockstamp: BlockStamp,
+        historical_strikes: dict[StrikesValidator, StrikesList],
+        strikes_per_frame: dict[Frame, dict[StrikesValidator, int]],
+        blockstamp: ReferenceBlockStamp,
     ) -> dict[StrikesValidator, StrikesList]:
-        out: dict[StrikesValidator, StrikesList] = {}
+        out = deepcopy(historical_strikes)
 
-        for (no_id, pubkey), raw_strikes in strikes.items():
-            # NOTE: Cleanup sequences like [0,0,0] since they don't bring any information.
-            if not sum(raw_strikes):
-                continue
-            maxlen = self.w3.csm.get_strikes_params(no_id, blockstamp).lifetime
-            if (no_id, pubkey) in out:
-                raise CSMError(f"Double accounting of strikes for {no_id=}")
-            out[(no_id, pubkey)] = StrikesList(raw_strikes, maxlen=maxlen)
+        for frame in self.state.frames:
+            strikes_in_frame = strikes_per_frame[frame]
+            for key in strikes_in_frame:
+                if key not in out:
+                    out[key] = StrikesList()
+                out[key].push(strikes_in_frame[key])
```

Comment:
`out: defaultdict` ?

Reply:
It can be, but I decided to keep this branch explicit.